app/scripts/controllers/emissionCtrl.js

```diff
@@ -294,117 +307,179 @@ var emissionCtrl = function($scope, $sce, walletService, libreService, $rootScop
+    $scope.updateRatesModal = function() {
+        setOpenModal(coeff.gasUpdateRates, $scope.updateRatesTx, "LIBRE_captionUpdateRates");
+    }
 
-        $scope.tx.data = txData;
-        $scope.tx.to = bankAddress;
-        $scope.tx.gasLimit = gasEmission;
-        $scope.tx.value = $scope.buyTXValue;
+    $scope.approveModal = function() {
+        setOpenModal(coeff.gasApprove, $scope.approveTx, "LIBRE_captionApprove");
+    }
 
-        libreTransaction($scope, "buyPending", "BUY", $translate, updateBalanceAndAllowance);
+    $scope.calcRatesModal = function() {
+        setOpenModal(coeff.gasCalcRates, $scope.calcRatesTx, "LIBRE_captionCalcRates");
     }
 
+    $scope.buyModal = function() {
+        setOpenModal(coeff.gasEmission, $scope.buyTx, "LIBRE_captionBuyTransaction");
+    }
 
-    $scope.generateApproveTx = function() {
-        approveTx();
+    $scope.sellModal = function() {
+        setOpenModal(coeff.gasRemission, $scope.sellTx, "LIBRE_captionSellTransaction");
     }
 
-    var approveTx = function() {
-        if (walletService.wallet == null) {
-            $translate("LIBRE_nullWallet").then((msg) => {
-                $scope.notifier.danger(msg);
-            })
-            return;
-        }
-        getTokenData("allowance", [walletService.wallet.getAddressString(), bankAddress]).then((allowanceData) => {
-            if (allowanceData.error) {
-                $scope.notifier.danger(allowanceData.msg);
-                return;
+    function prepareApproveTx() {
+        return new Promise((resolve, reject) => {
+            if (walletService.wallet == null) {
+                $translate("LIBRE_nullWallet").then((msg) => reject(msg));
             }
-            var allowance = +allowanceData.data[0];
+            getTokenData("allowance", [walletService.wallet.getAddressString(), bankAddress]).then((allowanceData) => {
+                if (allowanceData.error) {
+                    reject(allowanceData.msg);
+                }
+                var allowance = +allowanceData.data[0];
+                var tokensToAllowCount = $scope.tokensToAllow * Math.pow(10, libreService.coeff.tokenDecimals);
+                if (allowance == tokensToAllowCount) {
+                    $translate("LIBREALLOWANCE_equal").then((msg) => reject(msg));
+                } else if (tokensToAllowCount < allowance) {
+                    $scope.tx.data = getDataString(
+                        cashAbiRefactor["decreaseApproval"], [bankAddress, allowance - tokensToAllowCount]
+                    );
+                } else {
+                    // tokensToAllowCount > allowance
+                    $scope.tx.data = getDataString(
+                        cashAbiRefactor["increaseApproval"], [bankAddress, tokensToAllowCount - allowance]
+                    );
+                }
+                $scope.tx.to = cashAddress;
+                //$scope.tx.gasLimit = gasApprove;
+                $scope.tx.value = 0;
+                resolve();
```

Comment:
Maybe resolve $scope.tx?

Reply:
we don't need any var being resolved from here